### ISSUE TITLE
Traffic: Previously expanded messages do not get recreated correctly

### ISSUE THREAD

**Issue by mvglasow** [navit-gps/navit#1113]
Traffic messages which are read from `traffic.xml` and have been expanded (i.e. segments are cached in `traffic.xml`) do not seem to get recreated correctly. They show on the map (to be verified once more) but routing does not consider them as expected.

### To reproduce

1. Add the following to `traffic.xml` (telling it that the border crossings around Hrodna, Belarus into Poland and Lithuania are closed):
```
  <message id="hack:Raigardas-Prywalki" receive_time="2021-01-01T00:00:00Z" update_time="2021-01-01T00:00:00Z" expiration_time="2025-12-31T23:59:59Z">
    <location directionality="BOTH_DIRECTIONS">
      <from>+53.941841 +23.972212</from>
      <to>+53.941418 +23.969563</to>
    </location>
    <events>
      <event class="RESTRICTION" type="RESTRICTION_CLOSED">
      </event>
    </events>
  </message>
  <message id="hack:Kuznica-Bruzgi" receive_time="2021-01-01T00:00:00Z" update_time="2021-01-01T00:00:00Z" expiration_time="2025-12-31T23:59:59Z">
    <location directionality="BOTH_DIRECTIONS">
      <from>+53.521889 +23.660578</from>
      <to>+53.521423 +23.659136</to>
    </location>
    <events>
      <event class="RESTRICTION" type="RESTRICTION_CLOSED">
      </event>
    </events>
  </message>
```
2. Start Navit and calculate a route from Warsaw to Vilnius.

3. Stop navigation, close and restart Navit and calculate the same route again.

### Expected result:

A route that avoids Hrodna, probably crossing from Poland into Lithuania at Sejny (DK16)/Lazdijai (135).

### Actual result:

The first attempt produces the desired route (and stores the expanded segments for the messages in `traffic.xml`). However, the second attempt ignores the traffic messages and happily sends me through the closed border crossing into Belarus, through Hrodna, and back into Lithuania over another closed border crossing.

### Additional information:

* Map material may make a difference; this was reproduced with OSM data as of 2020-08-07 (retrieved from the Navit map server on 2020-08-27). Without traffic distortions Navit should assume the shortest route to go through Belarus.
* Deleting the expanded segments from `traffic.xml` causes Navit to calculate the correct route again. So it seems to have something to do with the pre-expanded segments in `traffic.xml`: they are either stored incorrectly or re-read incorrectly.
* With pre-expanded segments in `traffic.xml`, although routing seems to ignore the traffic distortion, it does get rendered on the map.

**Comment by mvglasow** [comment]
Also, stale pre-expanded segments (as would happen when the map is updated before the relaunch) result in proper behavior, likely because stale segment data gets discarded and the location gets expanded from scratch, as would be the case for a location without segment data.

**Comment by mvglasow** [comment]
One of the affected segments gets stored in `traffic.xml` as follows:

```
type=traffic_distortion id=0x169876,0x532f maxspeed=0 delay=2977 flags=0xfff00002
0x28ab4d 0x6d1dd2
0x28ab63 0x6d1ddd
0x28abb7 0x6d1df9
type=traffic_distortion id=0x169876,0x532f maxspeed=0 delay=2977 flags=0xfff00001
0x28ab4d 0x6d1dd2
0x28ab63 0x6d1ddd
0x28abb7 0x6d1df9
```

The road is a single-carriageway road (i.e. the way is open in both directions) and the location affects both directions, hence we are getting a traffic distortion item for each direction, each with the appropriate oneway flags (the two least significant bits).

When I examine the properties of the items in GTK GUI, I see:
* There is one item for the road, as well as two traffic distortion items – looks OK.
* Each item has three coordinates, so apparently we didn’t lose any points – looks OK.
* The access flags for the road are 0xfff00040, i.e. `AF_ALL|AF_MAXSPEED` – looks OK.
* `maxspeed` and `delay` are the correct values for both distortion items – looks OK.
* Flags are 0xfff00002, i.e. `AF_ALL|AF_ONEWAYREV` for BOTH traffic distortion items – this is not OK. Unless there is a bug in the GUI, it means one of the two traffic distortion items is getting recreated for the wrong direction.

For other items I see:
* Some ways have no flags.
* Sometimes Navit generates a pair of distortion items with flags 0xfff00000 (i.e. no oneway flag). So far this has only been observed for ways without flags, but not for all of them (there are pairings of ways without flags but traffic distortions with oneway flags set for opposite directions). Odd, but wouldn’t expect that to cause the issue we’re looking at here (if at all, it would cause Navit to avoid the way in both directions when the event affects only one).
* Pairs of traffic distortions which have opposite oneway flags in `traffic.xml` have the same flags in the GUI, either `AF_ONEWAY` or `AF_ONEWAYREV`.

Also, cached segments in `traffic.xml` are not changed when we exit for the second time. In particular, pairs of segments with opposite oneway flags retain these flags (the error shown in the GUI is not written back).

**Comment by mvglasow** [comment]
If I change the directionality to `ONE_DIRECTION`, all segments have oneway flags set.

Relaunching may alter segment data; I have noticed two differences here:
* Ordering of attributes (flags move to end)
* Delay values change: this may be caused because I am testing with a contradictory event combination of `DELAY_DELAY` and `RESTRICTION_CLOSED`.

Removing `DELAY_DELAY` from the message resulted in items with no `delay` attribute, and the only change after relaunch is the ordering of `maxspeed` and `flags`.

Even with the directionality changed to `ONE_DIRECTION` and `RESTRICTION_CLOSED` as the only event, the same behavior remains: correct when segments are expanded from scratch, but traffic distortion segments read from `traffic.xml` are ignored.

Examining freshly expanded versus pre-expanded segments in the GUI revealed no differences as far as I can tell. Specifically, the access flags seem correct.

Bottom line: Whether segments are expanded or re-read from cache, Navit seems to have the exact same data, save for a few oddities which we have eliminated from the test case.

One more thing to test: route recalculations after expanding/restoring segments. That is: start with no cached segments, calculate route, clear and recalculate, quit and relaunch, calculate route, clear and recalculate – which of the four attempts give the correct result?

**Comment by mvglasow** [comment]
Recalculated routes behave exactly the same as the previous route: traffic distortions are honored when segments are expanded from scratch, but not when read from `traffic.xml`. So, failure to add traffic distortions to the route graph is unlikely, as this should be rectified by a route recalculation.

**Comment by mvglasow** [comment]
Ran a test with Valgrind; the only thing that looked remotely related was a read after free error in `traffic_message_restore_segments()`. It was easy enough to fix (two statements to swap), but that still did not help with this error.

**Comment by mvglasow** [comment]
Dumping freshly generated traffic distortion items to the log resulted in the same data (semantically) which gets written to `traffic.xml`, only exception being that the dumped data does not contain the ID. Coordinates are the exact same as those for the corresponding ways.

In `traffic.c`, function `tm_item_add_message_data()`, removing the conditional before the call to `tm_item_update_attrs()` (forcing the call even where we thought we didn’t need it) also had no effect.

**Comment by mvglasow** [comment]
So, after various tests, it seems that segments restored from `traffic.xml` are semantically equivalent to freshly expanded ones. Also, I haven’t found any signs that the route is not being notified of their presence: the two places in the code where segments get expanded first try to restore segments from cache, and if that produces no map items, they expand from scratch. Both functions call the same two functions to add the item to the traffic map.

For all intents and purposes, in either case, the same segments with the same data get added to the map, and the routing engine gets notified in the same manner. Still, for some reason, the routing engine ignores segments restored from `traffic.xml`. Need to look at traffic distortion processing in the routing engine – it might be that restored segments represent their information somewhat differently and that triggers a hitherto undiscovered bug in the routing engine. Just a wild guess, the routing engine might assume attributes to be present in a certain order (i.e. fail to rewind between querying attributes) and act up if the order is not as expected.

**Comment by mvglasow** [comment]
I added some debug code to `route_get_traffic_distortion()` to produce a log message whenever a matching traffic distortion is found. When the segments are generated from scratch, I get a message for each segment, but for segments read from `traffic.xml` I get nothing. Additional debug code in the loop in which the segments get evaluated (generating a log message when the distortion item has the same ID as the segment) tells me that segments read from `traffic.xml` do not even get examined.

**So, although the items seem to be present with all data looking normal, apparently they do not get added to the route graph, hence `route_get_traffic_distortion()` cannot find them and does not honor them.**

Secondary finding: what is the meaning of `route_segment_data.len`? `route_get_traffic_distortion()` seems to equate it with the delay (presumably in deciseconds), but according to the doxygen comment it is the length in meters.

**Comment by mvglasow** [comment]
Adding extra debug messages to `route_graph_add_segment()` (with a conditional so they fire only for the ID of one particular segment) reveals that segments read from `traffic.xml` apparently do not get added to the route graph at all.

**Comment by mvglasow** [comment]
`tm_get_item()` gets called a few times while restoring segments (and returns data). However, when called (apparently) from `route_graph_build_idle()`, it returns null (i.e. no more items to return).

**Comment by mvglasow** [comment]
The calls while restoring segments appear to occur within `tm_add_item()`

**Comment by mvglasow** [comment]
Found one apparent error: when the list of items gets updated, `tm_rect_new()` may return a stale pointer to the first item. Fixing that, however, still results in an incorrect route. Need to dig deeper...

**Comment by mvglasow** [comment]
...and it turns out that, while `route_get_traffic_distortion()` does find the traffic distortion now, maxspeed is at `INT_MAX` when it should be zero. Looks like attributes do not get read correctly.

**Comment by mvglasow** [comment]
My gut feeling was right: attributes are not getting read correctly because the order is different for restored segments. `route_get_traffic_distortion()` does not set maxspeed because the appropriate flag is not set. That should happen in `route_graph_add_traffic_distortion()` when the item has a maxspeed attribute. However, the latter function reads three different attributes without calling `item_attr_rewind()` in between, which will only work correctly if the attributes happen to be stored in the same order in which they are read, but will fail if the order is different. Rewinding properly fixes that behavior, and maxspeed gets honored for restored segments.

Another thing I noticed is that delays cached in `traffic.xml` change when segments are restored and written back to `traffic.xml`. That is a different issue, though.

**Comment by mvglasow** [comment]
Still left to clean up, though not related to this issue:

* What is the meaning of `route_segment_data.len`? `route_get_traffic_distortion()` seems to equate it with the delay (presumably in deciseconds), but according to the doxygen comment it is the length in meters. Also check related structs.
* Delays cached in `traffic.xml` change when segments are restored and written back to `traffic.xml`.

**Comment by mvglasow** [comment]
Delays are fixed in #1116.
